### PR TITLE
Adding SAM CLI validation before SAM calls

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description
* Verifies that the SAM CLI is an acceptable version before running a SAM call
* Rechecks status only when the SAM CLI executable changes on disk (by modified time)
* Removed Command Palette command for verification in favor of an on-the-fly check

## Motivation and Context
We had no automatic validation of whether or not the SAM CLI will work

## Related Issue(s)

https://github.com/aws/aws-toolkit-vscode/issues/329

## Testing

Unit tests ran successfully. Deployment works successfully (don't have a version of the SAM CLI old enough to test whether or not the code works)

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by bryceitoc9** [comment]
n.b.: Might add a couple more tests in order to ensure updating the SAM CLI in-flight works.

File: src/test/shared/sam/cli/samCliInvoker.test.ts [hunk 1]
```diff
@@ -5,18 +5,83 @@
...
+import { TestLogger } from '../../../../shared/loggerUtils'
 import { SamCliConfiguration } from '../../../../shared/sam/cli/samCliConfiguration'
-import { DefaultSamCliProcessInvoker } from '../../../../shared/sam/cli/samCliInvoker'
+import {
+    DefaultSamCliProcessInvoker,
+    makeSamCliProcessInvokerContext
+} from '../../../../shared/sam/cli/samCliInvoker'
 import { assertRejects } from '../../utilities/assertUtils'
 
 describe('DefaultSamCliInvoker', async () => {
```
Comment (bryceitoc9) [inline src/test/shared/sam/cli/samCliInvoker.test.ts]: Realizing I don't have any happy path tests. Looking into it...

File: src/shared/sam/cli/samCliInvoker.ts [hunk 2]
```diff
@@ -8,42 +8,168 @@
...
+
+        if (validationResult.validation === SamCliVersionValidation.Valid) {
+            const args = typeof first === 'string' ? [ first, ...rest ] : rest
+            const options: SpawnOptions | undefined = typeof first === 'string' ? undefined : first
+
+            return await this.runCliCommand(samCliLocation, options, ...args)
         }
 
+        const errorResult: ChildProcessResult = {
+            exitCode: 1,
```
Comment (awschristou) [inline src/shared/sam/cli/samCliInvoker.ts]: Prefer a const instead of a magic number.

File: src/shared/sam/cli/samCliInvoker.ts [hunk 3]
```diff
@@ -8,42 +8,168 @@
...
+        if (exitCode === 0) {
+            const response = this.convertOutput(stdout)
+            if (!!response) {
+                return response
+            }
+
+            throw new Error('SAM CLI did not return expected data')
+        }
+
+        console.error('SAM CLI error')
```
Comment (awschristou) [inline src/shared/sam/cli/samCliInvoker.ts]: I realize this code was moved over, consider using the logger now that we have it.
If you choose to do this, please also log the command that was run.

File: src/shared/sam/cli/samCliInvoker.ts [hunk 4]
```diff
@@ -8,42 +8,168 @@
...
+        const validationResult = await this.getCliValidation(samCliLocation)
+
+        if (validationResult.validation === SamCliVersionValidation.Valid) {
+            const args = typeof first === 'string' ? [ first, ...rest ] : rest
+            const options: SpawnOptions | undefined = typeof first === 'string' ? undefined : first
+
+            return await this.runCliCommand(samCliLocation, options, ...args)
         }
 
+        const errorResult: ChildProcessResult = {
```
Comment (mpiroc) [inline src/shared/sam/cli/samCliInvoker.ts]: `ChildProcessResult` is intended to represent the raw results of a child process. So `exitCode` should be the child process's exit code, and `error` be any error thrown by running the child process.

Instead, suggest to use the default `Error` type, or your own extension of this type (possibly with a `validationResult` property).

File: src/shared/sam/cli/samCliInvoker.ts [hunk 5]
```diff
@@ -8,42 +8,168 @@
...
+        const samCliLocation = this._context.cliConfig.getSamCliLocation()
         if (!samCliLocation) {
-            throw new Error('SAM CLI location not configured')
+            const err = new Error('SAM CLI location not configured')
+            logger.error(err)
+            throw err
+        }
+        const validationResult = await this.getCliValidation(samCliLocation)
+
+        if (validationResult.validation === SamCliVersionValidation.Valid) {
```
Comment (mpiroc) [inline src/shared/sam/cli/samCliInvoker.ts]: IMO this logic belongs in the validator. The invoker should just call `validator.validate`, catch any error that it throws, and notify the user.

File: src/shared/sam/cli/samCliVersionValidator.ts [hunk 6]
```diff
@@ -35,46 +43,14 @@ export class SamCliVersionValidator {
...
-                'AWS.samcli.notification.version.valid',
-                'Your SAM CLI version {0} is valid.',
-                version
-            ),
-            SamCliVersionValidator.STATUS_BAR_DURATION_MILLIS
-        )
-    }
-
-    private async notifyVersionIsNotValid(validationResult: SamCliVersionValidatorResult): Promise<void> {
+    public async notifyVersionIsNotValid(validationResult: SamCliVersionValidatorResult): Promise<void> {
```
Comment (mpiroc) [inline src/shared/sam/cli/samCliVersionValidator.ts]: This doesn't feel like the right place to expose notification. The validator is responsible for determining what (if anything) is wrong, the caller is responsible for relaying that information to the user.

Consider either moving notification logic to the caller, or adding a new `Notifier` type to handle notification logic.

File: src/test/shared/sam/cli/samCliInvoker.test.ts [hunk 7]
```diff
@@ -5,18 +5,83 @@
...
+                stat: async () => testStat
+            },
+            cliInfo: { info: { version: testHighLevel }, lastModified: testDate }
+        })
+
+        const invoker = new DefaultSamCliProcessInvoker(context)
+
+        const result = await invoker.invoke()
+        assert.strictEqual(result.exitCode, 1)
+        assert.strictEqual(result.stderr, 'AWS Toolkit is out of date')
```
Comment (mpiroc) [inline src/test/shared/sam/cli/samCliInvoker.test.ts]: What if SAM updates, but we haven't published a new version of the toolkit yet? Should we instruct users to downgrade SAM CLI?

File: src/shared/sam/cli/samCliInvoker.ts [hunk 8]
```diff
@@ -8,42 +8,168 @@
...
+interface SamCliProcessInvokerContext {
+    cliConfig: SamCliConfiguration
+    cliInfo: SamCliProcessInfo
+    cliUtils: SamCliUtils
+    validator: SamCliVersionValidator
 }
 
-export interface SamCliTaskInvoker {
-    invoke(task: vscode.Task): Promise<vscode.TaskExecution>
+export function makeSamCliProcessInvokerContext(
```
[the remark this answers is not in the document]
Comment (mpiroc) [reply src/shared/sam/cli/samCliInvoker.ts]: Also re consistency: Elsewhere in our codebase, we define a class `DefaultFooContext`. Unless there's a good reason to avoid it, can you use that style?

File: src/shared/sam/cli/samCliInvoker.ts (hunk 5 above)
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: The problem is the `this.getCliValidation` command, since that would make another dependency loop.

File: src/shared/sam/cli/samCliInvoker.ts (hunk 3 above)
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: We log underneath. I'll go ahead and remove this (we also have this same setup in every other SAM CLI command, should I do the same to those as well?)

File: src/test/shared/sam/cli/samCliInvoker.test.ts (hunk 7 above)
Comment (bryceitoc9) [reply src/test/shared/sam/cli/samCliInvoker.test.ts]: Very good question. I have no idea! We should probably discuss this one at some point... We could maybe prompt the user to see if they want to give the command a shot?

File: src/shared/sam/cli/samCliInvoker.ts [hunk 9]
```diff
@@ -8,42 +8,168 @@
...
         return await childProcess.run()
     }
+
+    private async getCliValidation(samCliLocation: string): Promise<SamCliVersionValidatorResult> {
+        const cliStat = await this._context.cliUtils.stat(samCliLocation)
+        if (!this._context.cliInfo.lastModified || cliStat.mtime !== this._context.cliInfo.lastModified) {
+            this._context.cliInfo.lastModified = cliStat.mtime
+            this._context.cliInfo.info = await this.getInfo(samCliLocation)
+        }
+        if (!this._context.cliInfo.info) {
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: The idea here was to make as few calls to the SAM CLI as possible, but Typescript didn't like the possibility of leaving out the info. That being said, that should work, it's a weird enough case that a last modified date would be present without any info that this should probably be just rolled into the other change.

File: src/shared/sam/cli/samCliInvoker.ts (hunk 3 above)
Comment (awschristou) [reply src/shared/sam/cli/samCliInvoker.ts]: Where are they logged underneath?
The other commands can be adjusted in a separate PR.

Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: Err, logged underneath these lines of code. These messages (both the console.error statements and the error itself) were present before the initial logger changes, I just co-opted the error itself for the logger statement.
```
        const err =
            new Error(`sam --info encountered an error: ${error && error.message ? error.message : stderr || stdout}`)
        logger.error(err)
        throw err
```

File: src/shared/sam/cli/samCliInvoker.ts [hunk 10]
```diff
@@ -8,42 +8,168 @@
...
+    ): Promise<ChildProcessResult>  {
         const childProcess: ChildProcess = new ChildProcess(samCliLocation, options, ...args)
 
         return await childProcess.run()
     }
+
+    private async getCliValidation(samCliLocation: string): Promise<SamCliVersionValidatorResult> {
+        const cliStat = await this._context.cliUtils.stat(samCliLocation)
+        if (!this._context.cliInfo.lastModified || cliStat.mtime !== this._context.cliInfo.lastModified) {
+            this._context.cliInfo.lastModified = cliStat.mtime
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: `cliInfo.lastModified` is only set with a defined value if you're manually passing in a `cliInfo` object...the only case this should be happening in is test cases.

File: src/shared/sam/cli/samCliInvoker.ts (hunk 3 above)
Comment (awschristou) [reply src/shared/sam/cli/samCliInvoker.ts]: Ah okay. Would it be more useful to unconditionally log all of the fields? That statement has error.message potentially mask out stderr and stdout.

File: src/shared/sam/cli/samCliInvoker.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (awschristou) [reply src/shared/sam/cli/samCliInvoker.ts]: To rephrase, should setting lastModified happen after `this._context.cliInfo.info = await this.getInfo(samCliLocation)` ?

File: src/shared/sam/cli/samCliVersionValidator.ts (hunk 6 above)
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliVersionValidator.ts]: Honestly, if we remove this, we should probably just merge this file into the invoker (since the only other functionality is turning the version into an enum).

File: src/test/shared/sam/cli/samCliInvoker.test.ts (hunk 7 above)
Comment (bryceitoc9) [reply src/test/shared/sam/cli/samCliInvoker.test.ts]: Conversations with the JetBrains toolkit team indicate that we should break on advanced versions (it's better to favor safety in case the new SAM CLI has breaking changes).

File: src/shared/sam/cli/samCliInvoker.ts (hunk 3 above)
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: How's this?
```
        const err = new Error(
`sam --info encountered an error: ${error}
    ${error && error.message ? 'message: ' + error.message : ''}
    stderr : ${stderr}
    stdout: ${stdout}`
        )
```

File: src/shared/sam/cli/samCliInvoker.ts [hunk 11]
```diff
@@ -8,42 +8,168 @@
...
+        switch (validationResult.validation) {
+            case SamCliVersionValidation.VersionTooHigh:
+                const samVersionTooHighMessage = 'AWS Toolkit is out of date'
+                errorResult.error = new Error(samVersionTooHighMessage)
+                errorResult.stdout = samVersionTooHighMessage
+                errorResult.stderr = samVersionTooHighMessage
+                break
+            case SamCliVersionValidation.VersionTooLow:
+            case SamCliVersionValidation.VersionNotParseable:
+            default:
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: Removed altogether

File: src/shared/sam/cli/samCliInvoker.ts (hunk 2 above)
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: Removed altogether.

File: src/shared/sam/cli/samCliInvoker.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: Done

File: src/shared/sam/cli/samCliVersionValidator.ts (hunk 6 above)
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliVersionValidator.ts]: Haven't done the above yet but strongly considering it.

File: src/test/shared/sam/cli/samCliInvoker.test.ts (hunk 1 above)
Comment (bryceitoc9) [reply src/test/shared/sam/cli/samCliInvoker.test.ts]: Still TBD...

File: src/shared/sam/cli/samCliInvoker.ts [hunk 12]
```diff
@@ -31,79 +32,65 @@ interface SamCliProcessInvokerContext {
...
-                break
+        if (validationResult.validation !== SamCliVersionValidation.Valid) {
+            // prompt will redirect to external URL for updating. We do not need to wait on this.
+            // tslint:disable-next-line: no-floating-promises
+            this._context.validator.notifyVersionIsNotValid(validationResult)
+            const versionErr = new InvalidSamError(validationResult)
+            this._context.logger.error(
+                versionErr,
+                `${validationResult.validation}: version ${validationResult.version}`
+            )
```
Comment (awschristou) [inline src/shared/sam/cli/samCliInvoker.ts]: Did you mean to throw versionErr or somehow exit out, or do you intend to proceed to call runCliCommand?

File: src/shared/sam/cli/samCliInvoker.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (awschristou) [reply src/shared/sam/cli/samCliInvoker.ts]: Let me try again.

Should it be like this:
```
this._context.cliInfo.lastModified = cliStat.mtime
this._context.cliInfo.info = await this.getInfo(samCliLocation)
```
or like this:
```
this._context.cliInfo.info = await this.getInfo(samCliLocation)
this._context.cliInfo.lastModified = cliStat.mtime
```
In short: What if the call to getInfo errs?

File: src/shared/sam/cli/samCliInvoker.ts (hunk 12 above)
Comment (bryceitoc9) [reply src/shared/sam/cli/samCliInvoker.ts]: Yup, I missed it. Just pushed a fix (tests were breaking because of this)
